File: apps/client-cli/src/core/incoming-item-store.ts
import os from "node:os";
import path from "node:path";
import { promises as fs } from "node:fs";
import type { ClipboardPayload } from "@sharepaste/client-core";

const extensionForMime = (payload: ClipboardPayload): string => {
  const mime = payload.mime.toLowerCase();
  if (mime === "image/png") {
    return ".png";
  }
  if (mime === "image/jpeg") {
    return ".jpg";
  }
  if (mime === "image/gif") {
    return ".gif";
  }
  if (mime === "image/webp") {
    return ".webp";
  }
  if (mime === "application/pdf") {
    return ".pdf";
  }
  if (mime === "text/plain") {
    return ".txt";
  }
  return payload.type === "image" ? ".img" : ".bin";
};

export class IncomingItemStore {
  constructor(private readonly baseDir = path.join(os.homedir(), ".sharepaste", "received")) {}

  async materialize(payload: ClipboardPayload, plaintext: Uint8Array): Promise<string> {
    await fs.mkdir(this.baseDir, { recursive: true });
    const filename = `${payload.createdAtUnix}-${payload.itemId}${extensionForMime(payload)}`;
    const filePath = path.join(this.baseDir, filename);
    await fs.writeFile(filePath, Buffer.from(plaintext));
    return filePath;
  }
}

File: apps/mobile-android/app/src/main/kotlin/dev/sharepaste/android/data/IncomingItemStore.kt
package dev.sharepaste.android.data

import android.content.Context
import androidx.core.content.FileProvider
import dev.sharepaste.android.data.model.InboxItem
import dev.sharepaste.android.data.model.PayloadKind
import java.io.File

class IncomingItemStore(private val context: Context) {
    private val baseDir: File = File(context.filesDir, "received").apply { mkdirs() }

    fun materialize(
        itemId: String,
        kind: PayloadKind,
        mime: String,
        createdAtUnix: Long,
        sourceDeviceId: String,
        plaintext: ByteArray
    ): InboxItem {
        if (kind == PayloadKind.TEXT) {
            val preview = plaintext.toString(Charsets.UTF_8)
            return InboxItem(
                itemId = itemId,
                kind = kind,
                mime = mime,
                createdAtUnix = createdAtUnix,
                sourceDeviceId = sourceDeviceId,
                preview = preview
            )
        }

        val file = File(baseDir, "$createdAtUnix-$itemId${extensionForMime(mime, kind)}")
        file.writeBytes(plaintext)
        return InboxItem(
            itemId = itemId,
            kind = kind,
            mime = mime,
            createdAtUnix = createdAtUnix,
            sourceDeviceId = sourceDeviceId,
            filePath = file.absolutePath
        )
    }

    fun openUri(filePath: String) = FileProvider.getUriForFile(
        context,
        "${context.packageName}.files",
        File(filePath)
    )

    private fun extensionForMime(mime: String, kind: PayloadKind): String {
        return when (mime.lowercase()) {
            "image/png" -> ".png"
            "image/jpeg" -> ".jpg"
            "image/gif" -> ".gif"
            "image/webp" -> ".webp"
            "application/pdf" -> ".pdf"
            "text/plain" -> ".txt"
            else -> if (kind == PayloadKind.IMAGE) ".img" else ".bin"
        }
    }
}
